### PR TITLE
Template Variables: Expose getAdhocFilters and getVariableName

### PR DESCRIPTION
**Pull request by idastambuk**
**What is this feature?**

This PR exposes two more methods from TemplateSrv in grafana/runtime package. When[ using templateSrv in an external plugin](https://github.com/grafana/iot-sitewise-datasource/pull/172), I noticed it had access to all the methods from [TemplateSrv](https://github.com/grafana/grafana/blob/f8d89eff562dfb7c90b0bb617c639339af30d55c/public/app/features/templating/template_srv.ts), but when using them I got a type error. Not sure why other public methods weren't exposed, so I added just these two to the type because they are the most commonly used in internal plugins, so I assume external plugins are more likely to need them. Technically, though, I only need getVariableName for the Sitewise PR so this is up for discussion. 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
- [ ] There are no known compatibility issues with older supported versions of Grafana, or plugins.
- [ ] It passes the [Hosted Grafana feature readiness review](https://docs.google.com/document/d/1QL9Ly8KnXzpb6ISbg49pTODRO5mhA5tkkfIZVX6pqQU/edit) for observability, scalability, performance, and security.

### CODE REVIEW

File: packages/grafana-runtime/src/services/templateSrv.ts [hunk 1]
```diff
@@ -32,6 +32,16 @@ export interface TemplateSrv {
...
+
+  /**
+   * Lists ad-hoc filters for a specific datasource
+   */
+  getAdhocFilters(datasourceName: string): AdHocVariableFilter[];
+
+  /**
+   * Extracts variable name from an expression like '${example}'
+   */
+  getVariableName(expression: string): string | undefined | null;
```
Comment (matyax) [inline packages/grafana-runtime/src/services/templateSrv.ts:44]:
```suggestion
  getVariableName(expression: string): string | null;
```

What would be the use case for the `undefined` return value of this function? I mean, having 2 different falsy return types looks confusing at first glance.

**Review by matyax — COMMENTED** [review]
Loki changes look good, left a comment about a function.

File: packages/grafana-runtime/src/services/templateSrv.ts (hunk 1 above)
Comment (matyax) [reply packages/grafana-runtime/src/services/templateSrv.ts]: That's perfect, thank you!

File: packages/grafana-runtime/src/services/templateSrv.ts [hunk 2]
```diff
@@ -32,6 +32,16 @@ export interface TemplateSrv {
...
+
+  /**
+   * Lists ad-hoc filters for a specific datasource
+   */
+  getAdhocFilters(datasourceName: string): AdHocVariableFilter[];
+
+  /**
+   * Extracts variable name from an expression like '${example}'
+   */
+  getVariableName(expression: string): string | undefined;
```
Comment (torkelo) [inline packages/grafana-runtime/src/services/templateSrv.ts:44]: I don't think we want to expose these, not like this at least

Comment (mckn) [reply packages/grafana-runtime/src/services/templateSrv.ts:44]: I think we might need to expose something like this or an alternative approach to achieve this in a near future. Once we start to move core plugins out from the grafana/grafana repository we need to have a replacement.

**Review by mckn — COMMENTED** [review]
I guess we need to figure out how to expose this functionality to external plugins. @torkelo can you elaborate a bit more on your thoughts on this?

File: packages/grafana-runtime/src/services/templateSrv.ts (hunk 2 above)
Comment (torkelo) [reply packages/grafana-runtime/src/services/templateSrv.ts:44]: what is the use case for this function? 

We are doing lots of changes to templating at the moment

https://github.com/grafana/grafana/pull/65411
https://github.com/grafana/grafana/pull/65317 

as well as a new templating system in scenes :) 

So want limit any changes to the current system. 

Think  getAdhocFilters(datasourceName: string): should be fine, that is something some external plugins already use? 

 getVariableName(expression: string): string | undefined; is trickier, what about fieldPath, and formats? (in expressions like `${varName.fieldPath:format}`

Comment (idastambuk) [reply packages/grafana-runtime/src/services/templateSrv.ts:44]: I didn't know about any future plans to change this part of the templating service so I opened the PR to get some feedback. I'm open to alternative ideas, and/or just shelving this, until the improvements to the system are finished.

(The use case for getVariableName is [here](https://github.com/grafana/iot-sitewise-datasource/pull/172). 
I added getAdHocFilters as well when I looked at internal plugins and saw that they were using it and, as @mckn mentioned, we will be moving out those and they will need to be exposed in some way eventually.)

**Comment by idastambuk** [comment]
It doesn't look like we will be able to reach a consensus so Im closing this until the new templating system is in place

**Pull request closed without merging**